### PR TITLE
Sinc

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**

### CODE REVIEW

File: nutils/evaluable.py [hunk 1]
```diff
@@ -2330,6 +2330,14 @@ class ArcTan(Pointwise):
     return_type = lambda T: complex if T == complex else float
 
 
+class Sinc(Pointwise):
+    'Sine, element-wise.'
```
Comment (joostvanzwieten) [inline nutils/evaluable.py:2334]: Typo: 'sine' -> 'sinc'

File: nutils/numeric.py [hunk 2]
```diff
@@ -591,4 +591,35 @@ def levicivita(n: int, dtype=float):
...
+    # NOTE: alternatively we could initialize f with zeros and iterate from i=0
+    # below, but we prefer to benefit from the numerical stability that Numpy's
+    # implementation provides
+    for i in range(1, n+1):
+        f *= -i
+        f += numpy.sin(x + i * numpy.pi / 2) # derivative corresponds to ½π shift
+        with numpy.errstate(all='ignore'): # results for x=0 are overwitten below
+            f /= x
+    if n and 0 in x:
+        f[x == 0] = [1, 0, -1, 0][n % 4] / (n + 1)
```
Comment (joostvanzwieten) [inline nutils/numeric.py:621]: This doesn't work if `f` is scalar:

```python
>>> nutils.numeric.sinc(numpy.array(0.), n=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/joost/projects/nutils/nutils/nutils/numeric.py", line 621, in sinc
    f[x == 0] = [1, 0, -1, 0][n % 4] / (n + 1)
    ~^^^^^^^^
TypeError: 'numpy.float64' object does not support item assignment
```

File: nutils/numeric.py [hunk 3]
```diff
@@ -591,4 +591,35 @@ def levicivita(n: int, dtype=float):
...
+
+    f = numpy.sinc(x / numpy.pi) # Numpy's implementation is normalized to π
+    # NOTE: alternatively we could initialize f with zeros and iterate from i=0
+    # below, but we prefer to benefit from the numerical stability that Numpy's
+    # implementation provides
+    for i in range(1, n+1):
+        f *= -i
+        f += numpy.sin(x + i * numpy.pi / 2) # derivative corresponds to ½π shift
+        with numpy.errstate(all='ignore'): # results for x=0 are overwitten below
+            f /= x
```
Comment (joostvanzwieten) [inline nutils/numeric.py:619]: The derivative is not stable in the neighborhood of 0:

```python
>>> nutils.numeric.sinc(numpy.array([1e-6, 1e-7, 1e-8, 1e-9, 0]), n=2)
array([-0.33320005, -0.33386618, -0.99999998, -0.99999996, -0.33333333])
```

**Review by joostvanzwieten — CHANGES_REQUESTED** [review]

File: nutils/evaluable.py (hunk 1 above)
Comment (gertjanvanzwieten) [reply nutils/evaluable.py]: fixed (removed docstring)

File: nutils/numeric.py (hunk 2 above)
Comment (gertjanvanzwieten) [reply nutils/numeric.py]: fixed

File: nutils/numeric.py (hunk 3 above)
Comment (gertjanvanzwieten) [reply nutils/numeric.py:620]: fixed

**Review by joostvanzwieten — DISMISSED** [review]

File: nutils/numeric.py [hunk 4]
```diff
@@ -591,4 +591,51 @@ def levicivita(n: int, dtype=float):
...
+        sinc(x) = sin(x) / x
+    '''
+
+    x = numpy.asarray(x)
+    f = numpy.sinc(x / numpy.pi) # Numpy's implementation is normalized to π
+    if n == 0:
+        return f
+    # Derivatives are evaluated using either a recurrence relation or a Taylor
+    # series expansion, depending on proximity to the origin.
+    m = x >= 1
```
Comment (joostvanzwieten) [inline nutils/numeric.py:608]: Shouldn't this be `abs(x) >= 1`?

File: nutils/numeric.py [hunk 5]
```diff
@@ -591,4 +591,51 @@ def levicivita(n: int, dtype=float):
...
+    if m.any(): # points outside unit ball
+        # sinc'i(x) = (sin'i(x) - i sinc'i-1(x)) / x
+        fm = f[m]
+        xm = x[m]
+        for i in range(1, n+1): # traverse derivatives
+            fm *= -i
+            fm += [numpy.sin, numpy.cos][i%2](xm) * [1, 1, -1, -1][i % 4]
+            fm /= xm
+        if not f.shape:
+            return float(fm)
```
Comment (joostvanzwieten) [inline nutils/numeric.py:618]: The cast to `float` is wrong if `x` is complex. Besides the early return, is there a reason for this branch?

**Review by joostvanzwieten — CHANGES_REQUESTED** [review]

File: nutils/numeric.py (hunk 5 above)
Comment (gertjanvanzwieten) [reply nutils/numeric.py]: fixed

**Review by joostvanzwieten — APPROVED** [review]

**Pull request merged**